### PR TITLE
DEPR: _BaseOffset's parameter `n` should be prevented from being 0

### LINKED ISSUE #31184: _BaseOffset's parameter `n` should be prevented from being 0
The docstring says
```
        Require that `n` be a nonzero integer.
```
but the "nonzero" condition isn't enforced.

Enforcing it would prevent the undesired behaviour noticed in #29901 , whereby
```
cbh = pd.offsets.CustomBusinessHour(start='07:00', end='01:00')
time = pd.Timestamp('2019-03-06 13:00:00')
time + 0*cbh
```
was returning
```
2019-03-07 07:00:00
```

### PR DESCRIPTION
**Pull request by MarcoGorelli**
- [ ] closes #31184 
- [ ] tests added / passed
- [ ] passes `black pandas`
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

**Comment by WillAyd** [comment]
I think @jbrockmendel has done similar work on these in the past

File: doc/source/whatsnew/v1.1.0.rst [hunk 1]
```diff
@@ -94,7 +94,9 @@ Datetimelike
 ^^^^^^^^^^^^
 - Bug in :class:`Timestamp` where constructing :class:`Timestamp` from ambiguous epoch time and calling constructor again changed :meth:`Timestamp.value` property (:issue:`24329`)
 - :meth:`DatetimeArray.searchsorted`, :meth:`TimedeltaArray.searchsorted`, :meth:`PeriodArray.searchsorted` not recognizing non-pandas scalars and incorrectly raising ``ValueError`` instead of ``TypeError`` (:issue:`30950`)
--
+- Date offsets can no longer be constructed with multiplicity 0, e.g. ``dt.datetime(2019, 3, 6, 13, 0) + 0*pd.offsets.MonthEnd()`` will raise
```
Comment (jbrockmendel) [inline doc/source/whatsnew/v1.1.0.rst]: multiplicity -> `n=0`

**Comment by jbrockmendel** [comment]
Are we sure this should be enforced for _all_ (non-tick) DateOffset subclasses?  `Week` comes to mind as a possible exception.

**Comment by Liam3851** [comment]
@jbrockmendel As I noted on the issue, n=0 is currently docced as a rollforward for the general DateOffset class, although it's suggested that "since 0 is a bit weird, we suggest avoiding its use". Perhaps the first step should be officially deprecating n=0 DateOffsets before removal?

With Week, so far as I can tell Week(n=0) does absolutely nothing (not even a rollforward). Not sure whether that speaks in its favor (doing nothing for a 0 offset is unambiguous and in some sense logical) or against it (using it doesn't actually do anything)?

```python
In [3]: pd.to_datetime('2020-01-22') + pd.offsets.Week(n=0)
Out[3]: Timestamp('2020-01-22 00:00:00')

In [4]: pd.to_datetime('2020-01-22') + pd.offsets.Week(n=1)
Out[4]: Timestamp('2020-01-29 00:00:00')

In [5]: pd.to_datetime('2020-01-21') + pd.offsets.Week(n=0)
Out[5]: Timestamp('2020-01-21 00:00:00')

In [6]: pd.to_datetime('2020-01-21 01:29') + pd.offsets.Week(n=0)
Out[6]: Timestamp('2020-01-21 01:29:00')
```

**Comment by jbrockmendel** [comment]
> although it's suggested that "since 0 is a bit weird, we suggest avoiding its use". Perhaps the first step should be officially deprecating n=0 DateOffsets before removal?

I think this makes sense, cc @jreback @mroeschke 

> With Week, so far as I can tell Week(n=0) does absolutely nothing (not even a rollforward).

`Week(n=whatever, normalize=False, weekday=None)` is equivalent to `Day(n=7*whatever)`.  It's kind of annoying to have that one tick-like non-Tick.  Could deprecate the weekday=None option (I'm getting off topic)

**Comment by MarcoGorelli** [comment]
@jbrockmendel Sorry, having some trouble with this one.

I've added a FutureWarning to `_validate_n` from `pandas/_libs/tslibs/offsets.pyx::_BaseOffset`, re-ran
```
$ python setup.py develop
```
and then added the decorator
```
    @pytest.mark.filterwarnings("ignore::FutureWarning")
```
to `_check_offsetfunc_works` from `pandas/tests/tseries/offsets/test_offsets.py::TestCommon`.

However, when I run
```
$ pytest pandas/tests/tseries/offsets/test_offsets.py::TestCommon::test_apply
```
I still get
```
AssertionError: Caused unexpected warning(s): [('FutureWarning', FutureWarning('Deprecated n=0'), '/home/SERILOCAL/m.gorelli/pandas/pandas/tseries/offsets.py', 1290)]
```

(note: I'm aware that "Deprecated n=0" isn't an adequate warning message, I just wanted to try getting this working before looking into how to write the message properly)

**Comment by jbrockmendel** [comment]
> @pytest.mark.filterwarnings("ignore::FutureWarning")

Instead, use `with tm.assert_produces_warning(FutureWarning):`

**Comment by MarcoGorelli** [comment]
@jbrockmendel sorry to tag you here again, just don't know what to make of the CI failure. If I run 
```
$ python setup.py develop
$ pytest pandas/tests/arithmetic/test_numeric.py
```
locally on Linux, Python 3.7.6, it passes. But during the "Linux py37_np_dev" CI job, this test fails.

Any suggestions on why this may be would be appreciated, else I'll keep trying to figure it out :)

> Instead, use with tm.assert_produces_warning(FutureWarning):

Just a note on this: the part of the code where I wanted to apply this (pandas/tests/tseries/offsets/test_offsets.py, line 367) doesn't always produce a ``FutureWarning``, so I wanted to just ignore the warning. Have used ``warnings.filterwarnings``

**Comment by jbrockmendel** [comment]
It looks like the n!=0 restriction is being applied to offsets.Second, when it shouldn't apply to any of the Tick offsets.

File: pandas/tests/tseries/offsets/test_offsets.py [hunk 2]
```diff
@@ -359,7 +360,11 @@ def _check_offsetfunc_works(self, offset, funcname, dt, expected, normalize=Fals
 
         # test nanosecond is preserved
         with tm.assert_produces_warning(exp_warning, check_stacklevel=False):
-            result = func(ts)
+            with warnings.catch_warnings():
+                warnings.filterwarnings(
+                    "ignore", ".*deprecated.*", category=FutureWarning
+                )
+                result = func(ts)
```
Comment (MarcoGorelli) [inline pandas/tests/tseries/offsets/test_offsets.py]: I'm not sure it's correct to filter warnings here

File: pandas/tests/tseries/offsets/test_offsets.py [hunk 3]
```diff
@@ -396,13 +401,31 @@ def _check_offsetfunc_works(self, offset, funcname, dt, expected, normalize=Fals
 
             # test nanosecond is preserved
             with tm.assert_produces_warning(exp_warning, check_stacklevel=False):
-                result = func(ts)
+                with warnings.catch_warnings():
+                    warnings.filterwarnings(
+                        "ignore", ".*deprecated.*", category=FutureWarning
+                    )
+                    result = func(ts)
```
Comment (MarcoGorelli) [inline pandas/tests/tseries/offsets/test_offsets.py]: not sure it's correct to filter warnings here

File: pandas/tests/arithmetic/test_period.py [hunk 4]
```diff
@@ -1251,11 +1251,11 @@ def test_pi_ops(self):
 
         result = idx - Period("2011-01", freq="M")
         off = idx.freq
-        exp = pd.Index([0 * off, 1 * off, 2 * off, 3 * off], name="idx")
+        exp = pd.Index([off, 1 * off, 2 * off, 3 * off], name="idx")
```
Comment (MarcoGorelli) [inline pandas/tests/arithmetic/test_period.py]: I don't think this should be changed, instead just check warning is produced

**Comment by MarcoGorelli** [comment]
@jbrockmendel - thanks, seems this is passing now. The [travis job](https://travis-ci.org/pandas-dev/pandas/builds/647315576?utm_source=github_status&utm_medium=notification) says it's passed if I click on it but still shows as pending here

**Review by jreback — CHANGES_REQUESTED** [review]
this is removing an awful lot of tests. can you comment on some of the removals why we no longer test these

**Comment by MarcoGorelli** [comment]
> this is removing an awful lot of tests. can you comment on some of the removals why we no longer test these

@jreback Some of these, e.g.
```
    apply_cases.append(
        (
            BDay(0),
            {
                datetime(2008, 1, 1): datetime(2008, 1, 1),
                datetime(2008, 1, 4): datetime(2008, 1, 4),
                datetime(2008, 1, 5): datetime(2008, 1, 7),
                datetime(2008, 1, 6): datetime(2008, 1, 7),
                datetime(2008, 1, 7): datetime(2008, 1, 7),
            },
        )
    )
```
cause a warning error during test collection. I can catch the warning / rewrite the expected value if this happens in the middle of a test, but don't know of a way to do that during test collection.

**Comment by jbrockmendel** [comment]
side-note: we should look into restricting FY5253 startingMOnth/weekday to be non-None

File: pandas/_libs/tslibs/offsets.pyx [hunk 5]
```diff
@@ -1,6 +1,7 @@
 import cython
 
 import time
+import warnings
```
Comment (jbrockmendel) [inline pandas/_libs/tslibs/offsets.pyx:4]: warnings is already imported on L5

File: pandas/_libs/tslibs/offsets.pyx [hunk 6]
```diff
@@ -533,6 +534,25 @@ class _BaseOffset:
             raise TypeError(f'`n` argument must be an integer, got {type(n)}')
         if n != nint:
             raise ValueError(f'`n` argument must be an integer, got {n}')
+        if n == 0:
+            if hasattr(self, "_prefix") and self._prefix in [
```
Comment (jbrockmendel) [inline pandas/_libs/tslibs/offsets.pyx:538]: We should probably make this function into a classmethod and avoid referencing "self" here

File: pandas/tests/arithmetic/test_datetime64.py [hunk 7]
```diff
@@ -1354,7 +1357,23 @@ def test_dt64arr_add_sub_DateOffsets(
 
         offset_cls = getattr(pd.offsets, cls_name)
 
-        with warnings.catch_warnings(record=True):
+        if PY37:
```
Comment (jbrockmendel) [inline pandas/tests/arithmetic/test_datetime64.py:1360]: maybe do this in compat or tm?

File: pandas/tests/arithmetic/test_period.py [hunk 8]
```diff
@@ -1276,13 +1276,15 @@ def test_pi_ops(self):
 
         self._check(idx + 2, lambda x: x - 2, idx)
 
-        result = idx - Period("2011-01", freq="M")
-        off = idx.freq
-        exp = pd.Index([0 * off, 1 * off, 2 * off, 3 * off], name="idx")
+        with tm.assert_produces_warning(FutureWarning, check_stacklevel=False):
+            result = idx - Period("2011-01", freq="M")
+            off = idx.freq
+            exp = pd.Index([0 * off, 1 * off, 2 * off, 3 * off], name="idx")
```
Comment (jbrockmendel) [inline pandas/tests/arithmetic/test_period.py:1282]: once this deprecation is enforced, what would happen here?  just not-allowing subtraction seems undesirable

Comment (MarcoGorelli) [reply pandas/tests/arithmetic/test_period.py:1282]: Yes, true that it seems undesirable. Any suggestions? I thought of returning an offset that allows for `n=0`, but that could be undesirable if other time arithmetic operations are involved.

For reference, when we get here, we have
```python
ipdb> idx                                                                                                                                     
PeriodIndex(['2011-01', '2011-02', '2011-03', '2011-04'], dtype='period[M]', name='idx', freq='M')
```

Comment (jbrockmendel) [reply pandas/tests/arithmetic/test_period.py:1282]: Looking back at #31184 wouldnt removing the incorrect doc also be a viable option?

Comment (MarcoGorelli) [reply pandas/tests/arithmetic/test_period.py:1282]: LOL yeah, that would be a lot simpler :) Perhaps there could still be a warning that an offset with `n=0` has been created - what do you think?

Comment (jbrockmendel) [reply pandas/tests/arithmetic/test_period.py:1282]: could the warning go closer to the problematic behavior?

**Pull request closed without merging**